File: scripts/usage-report.py
#!/usr/bin/env python3
"""Aggregate token/cost Usage events from <client-root>/audit/*.jsonl.
Usage: usage-report.py <client-root> [--by session|day|model|skill]   (default: session)
  --by skill joins sessions to decisions/<skill>/*.md via the `session:` frontmatter line, giving
  cost per task and per domain — the number a stress test or a billing review wants."""
import sys, json, glob, os, re, collections
root = sys.argv[1]; by = sys.argv[sys.argv.index("--by") + 1] if "--by" in sys.argv else "session"
events = []
for f in sorted(glob.glob(os.path.join(root, "audit", "*.jsonl"))):
    day = os.path.basename(f)[:-6]
    for line in open(f, encoding="utf-8"):
        try: e = json.loads(line)
        except Exception: continue
        e["_day"] = day; events.append(e)
usage = [e for e in events if e.get("event") == "Usage"]
if not usage: print("no Usage events"); sys.exit(0)

# session -> skill(s) via decision records
sess_skill = collections.defaultdict(set)
for f in glob.glob(os.path.join(root, "decisions", "*", "*.md")):
    skill = os.path.basename(os.path.dirname(f))
    for line in open(f, encoding="utf-8"):
        m = re.match(r"session:\s*(\S+)", line)
        if m: sess_skill[m.group(1)].add(skill); break

agg = collections.defaultdict(lambda: dict(turns=0, input=0, output=0, cache_write=0, cache_read=0, cost=0.0))
for e in usage:
    keys = {"session": [e.get("session", "")], "day": [e["_day"]],
            "model": list(e.get("turn_tokens", {}).keys()) or ["?"],
            "skill": sorted(sess_skill.get(e.get("session", ""), {"(no decision record)"}))}[by]
    for k in keys:
        a = agg[k]; a["turns"] += 1; a["cost"] += e.get("turn_cost_usd", 0.0)
        for model, t in e.get("turn_tokens", {}).items():
            if by == "model" and model != k: continue
            for kk in ("input", "output", "cache_write", "cache_read"): a[kk] += t.get(kk, 0)
w = max(len(str(k)) for k in agg) + 2
print(f"{by:<{w}}{'turns':>6}{'input':>10}{'output':>9}{'cache_w':>9}{'cache_r':>10}{'cost $':>9}")
tot = dict(turns=0, input=0, output=0, cache_write=0, cache_read=0, cost=0.0)
for k, a in sorted(agg.items()):
    print(f"{str(k):<{w}}{a['turns']:>6}{a['input']:>10}{a['output']:>9}{a['cache_write']:>9}{a['cache_read']:>10}{a['cost']:>9.2f}")
    for kk in tot: tot[kk] += a[kk]
print(f"{'TOTAL':<{w}}{tot['turns']:>6}{tot['input']:>10}{tot['output']:>9}{tot['cache_write']:>9}{tot['cache_read']:>10}{tot['cost']:>9.2f}")
chk = sorted({e.get("pricing_checked", "?") for e in usage}); print(f"\npricing table checked: {', '.join(chk)}  (estimates; subagent turns not included)")

File: scripts/usage-report.js
#!/usr/bin/env node
// Aggregate Usage events from <client-root>/audit/*.jsonl.
//   usage: node usage-report.js <client-root> [--by session|day|model|skill]
//   --by skill joins sessions to decisions/<skill>/*.md via the `session:` frontmatter line.
"use strict";
const fs = require("fs");
const path = require("path");
const { walk } = require("./lib.js");

const root = process.argv[2]; const by = process.argv.includes("--by") ? process.argv[process.argv.indexOf("--by") + 1] : "session";
const events = [];
const adir = path.join(root, "audit");
for (const f of (fs.existsSync(adir) ? fs.readdirSync(adir) : []).filter((n) => n.endsWith(".jsonl")).sort())
  for (const line of fs.readFileSync(path.join(adir, f), "utf8").split(/\r?\n/)) { try { const e = JSON.parse(line); e._day = f.slice(0, -6); events.push(e); } catch {} }
const usage = events.filter((e) => e.event === "Usage");
if (!usage.length) { console.log("no Usage events"); process.exit(0); }
const sessSkill = {};
for (const f of walk(path.join(root, "decisions"), (p) => p.endsWith(".md"))) {
  const m = fs.readFileSync(f, "utf8").match(/^session:\s*(\S+)/m);
  if (m) (sessSkill[m[1]] = sessSkill[m[1]] || new Set()).add(path.basename(path.dirname(f)));
}
const KEYS = ["input", "output", "cache_write", "cache_read"];
const agg = {}; const row = () => ({ turns: 0, input: 0, output: 0, cache_write: 0, cache_read: 0, cost: 0 });
for (const e of usage) {
  const keys = { session: [e.session || ""], day: [e._day], model: Object.keys(e.turn_tokens || {}).length ? Object.keys(e.turn_tokens) : ["?"],
    skill: [...(sessSkill[e.session] || new Set(["(no decision record)"]))].sort() }[by];
  for (const k of keys) {
    const a = agg[k] || (agg[k] = row()); a.turns++; a.cost += e.turn_cost_usd || 0;
    for (const [model, t] of Object.entries(e.turn_tokens || {})) { if (by === "model" && model !== k) continue; for (const kk of KEYS) a[kk] += t[kk] || 0; }
  }
}
const w = Math.max(...Object.keys(agg).map((k) => k.length)) + 2;
const pad = (s, n, r) => (r ? String(s).padStart(n) : String(s).padEnd(n));
console.log(pad(by, w) + pad("turns", 6, 1) + pad("input", 10, 1) + pad("output", 9, 1) + pad("cache_w", 9, 1) + pad("cache_r", 10, 1) + pad("cost $", 9, 1));
const tot = row();
for (const k of Object.keys(agg).sort()) { const a = agg[k]; console.log(pad(k, w) + pad(a.turns, 6, 1) + pad(a.input, 10, 1) + pad(a.output, 9, 1) + pad(a.cache_write, 9, 1) + pad(a.cache_read, 10, 1) + pad(a.cost.toFixed(2), 9, 1)); for (const kk of Object.keys(tot)) tot[kk] += a[kk]; }
console.log(pad("TOTAL", w) + pad(tot.turns, 6, 1) + pad(tot.input, 10, 1) + pad(tot.output, 9, 1) + pad(tot.cache_write, 9, 1) + pad(tot.cache_read, 10, 1) + pad(tot.cost.toFixed(2), 9, 1));
console.log(`\npricing table checked: ${[...new Set(usage.map((e) => e.pricing_checked || "?"))].sort().join(", ")}  (estimates; subagent turns not included)`);
